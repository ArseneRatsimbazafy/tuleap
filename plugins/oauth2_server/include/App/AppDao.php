<?php
/**
 * Copyright (c) Enalean, 2020-Present. All Rights Reserved.
 *
 * This file is a part of Tuleap.
 *
 * Tuleap is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * Tuleap is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Tuleap. If not, see <http://www.gnu.org/licenses/>.
 */

declare(strict_types=1);

namespace Tuleap\OAuth2Server\App;

use Tuleap\DB\DataAccessObject;

class AppDao extends DataAccessObject
{
    /**
     * @psalm-return array{id:int, project_id:int, name:string, redirect_endpoint: string}
     */
    public function searchByClientId(ClientIdentifier $client_id): ?array
    {
        $sql = 'SELECT * FROM plugin_oauth2_server_app
            WHERE id = ?';
        return $this->getDB()->row($sql, $client_id->getInternalId());
    }

    /**
     * @psalm-return array<array{id:int, project_id:int, name:string, redirect_endpoint: string}>
     */
    public function searchByProject(\Project $project): array
    {
        $sql = "SELECT * FROM plugin_oauth2_server_app
            WHERE project_id = ?";
        return $this->getDB()->run($sql, $project->getID());
    }

    public function create(NewOAuth2App $app): void
    {
        $this->getDB()->insert(
            'plugin_oauth2_server_app',
            [
                'project_id'        => $app->getProject()->getID(),
                'name'              => $app->getName(),
                'redirect_endpoint' => $app->getRedirectEndpoint()
            ]
        );
    }

    public function delete(int $app_id): void
    {
        $this->getDB()->run('DELETE FROM plugin_oauth2_server_app WHERE id = ?', $app_id);
    }
}
