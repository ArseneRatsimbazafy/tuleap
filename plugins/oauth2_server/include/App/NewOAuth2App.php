<?php
/**
 * Copyright (c) Enalean, 2020-Present. All Rights Reserved.
 *
 * This file is a part of Tuleap.
 *
 * Tuleap is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * Tuleap is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Tuleap. If not, see <http://www.gnu.org/licenses/>.
 */

declare(strict_types=1);

namespace Tuleap\OAuth2Server\App;

use Tuleap\Authentication\SplitToken\SplitTokenVerificationString;
use Tuleap\Authentication\SplitToken\SplitTokenVerificationStringHasher;

/**
 * @psalm-immutable
 */
final class NewOAuth2App
{
    /**
     * @var string
     */
    private $name;
    /**
     * @var string
     */
    private $redirect_endpoint;
    /**
     * @var SplitTokenVerificationString
     */
    private $secret;
    /**
     * @var string
     */
    private $hashed_secret;
    /**
     * @var \Project
     */
    private $project;

    private function __construct(
        string $name,
        string $redirect_endpoint,
        SplitTokenVerificationString $secret,
        string $hashed_secret,
        \Project $project
    ) {
        $this->name              = $name;
        $this->redirect_endpoint = $redirect_endpoint;
        $this->secret            = $secret;
        $this->hashed_secret     = $hashed_secret;
        $this->project           = $project;
    }

    /**
     * @throws InvalidAppDataException
     */
    public static function fromAppData(
        string $name,
        string $redirect_endpoint,
        \Project $project,
        SplitTokenVerificationStringHasher $hasher
    ): self {
        $is_data_valid = self::isAppDataValid($name, $redirect_endpoint);

        if (! $is_data_valid) {
            throw new InvalidAppDataException();
        }

        $secret = SplitTokenVerificationString::generateNewSplitTokenVerificationString();

        return new self(
            $name,
            $redirect_endpoint,
            $secret,
            $hasher->computeHash($secret),
            $project
        );
    }

    private static function isAppDataValid(string $name, string $redirect_endpoint): bool
    {
        $string_validator = new \Valid_String();
        $string_validator->required();
        // See https://tools.ietf.org/html/rfc6749#section-3.1.2
        $redirect_endpoint_validator = new \Valid_String();
        $redirect_endpoint_validator->required();
        $redirect_endpoint_validator->addRule(new \Rule_Regexp('/^https:\/\/[^#]*$/i'));

        return $string_validator->validate($name) && $redirect_endpoint_validator->validate($redirect_endpoint);
    }

    public function getName(): string
    {
        return $this->name;
    }

    public function getRedirectEndpoint(): string
    {
        return $this->redirect_endpoint;
    }

    public function getProject(): \Project
    {
        return $this->project;
    }

    public function getSecret(): SplitTokenVerificationString
    {
        return $this->secret;
    }

    public function getHashedSecret(): string
    {
        return $this->hashed_secret;
    }
}
