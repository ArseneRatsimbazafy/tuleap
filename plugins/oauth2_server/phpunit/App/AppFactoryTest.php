<?php
/**
 * Copyright (c) Enalean, 2020-Present. All Rights Reserved.
 *
 * This file is a part of Tuleap.
 *
 * Tuleap is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * Tuleap is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Tuleap. If not, see <http://www.gnu.org/licenses/>.
 */

declare(strict_types=1);

namespace Tuleap\OAuth2Server\App;

use Mockery as M;
use Mockery\Adapter\Phpunit\MockeryPHPUnitIntegration;
use PHPUnit\Framework\TestCase;

final class AppFactoryTest extends TestCase
{
    use MockeryPHPUnitIntegration;

    /**
     * @var AppFactory
     */
    private $app_factory;
    /**
     * @var M\LegacyMockInterface|M\MockInterface|AppDao
     */
    private $app_dao;

    protected function setUp(): void
    {
        $this->app_dao     = M::mock(AppDao::class);
        $this->app_factory = new AppFactory($this->app_dao);
    }

    public function testGetAppsForProject(): void
    {
        $rows    = [
            ['id' => 1, 'name' => 'Jenkins'],
            ['id' => 2, 'name' => 'My custom REST client']
        ];
        $project = M::mock(\Project::class);
        $this->app_dao->shouldReceive('searchByProject')
            ->once()
            ->with($project)
            ->andReturn($rows);

        $result = $this->app_factory->getAppsForProject($project);
        $this->assertEquals(
            [new OAuth2App(1, 'Jenkins', $project), new OAuth2App(2, 'My custom REST client', $project)],
            $result
        );
    }
}
