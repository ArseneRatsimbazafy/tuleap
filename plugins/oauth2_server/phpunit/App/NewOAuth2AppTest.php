<?php
/**
 * Copyright (c) Enalean, 2020-Present. All Rights Reserved.
 *
 * This file is a part of Tuleap.
 *
 * Tuleap is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * Tuleap is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Tuleap. If not, see <http://www.gnu.org/licenses/>.
 */

declare(strict_types=1);

namespace Tuleap\OAuth2Server\App;

use Mockery as M;
use Mockery\Adapter\Phpunit\MockeryPHPUnitIntegration;
use PHPUnit\Framework\TestCase;

final class NewOAuth2AppTest extends TestCase
{
    use MockeryPHPUnitIntegration;

    public function testFromAppDataThrowsWhenAppNameIsEmpty(): void
    {
        $this->expectException(InvalidAppDataException::class);
        NewOAuth2App::fromAppData('', 'https://example.com/redirect', M::mock(\Project::class));
    }

    public function testFromAppDataThrowsWhenRedirectEndpointIsEmpty(): void
    {
        $this->expectException(InvalidAppDataException::class);
        NewOAuth2App::fromAppData('Jenkins', '', M::mock(\Project::class));
    }

    public function testFromAppDataThrowsWhenRedirectEndpointIsNotHTTPS(): void
    {
        $this->expectException(InvalidAppDataException::class);
        NewOAuth2App::fromAppData('Jenkins', 'http://insecure.example.com', M::mock(\Project::class));
    }

    public function testFromAppDataThrowsWhenRedirectEndpointContainsAnAnchor(): void
    {
        $this->expectException(InvalidAppDataException::class);
        NewOAuth2App::fromAppData('Jenkins', 'https://example.com/redirect#fragment', M::mock(\Project::class));
    }

    public function testFromAppDataReturnsANewOauth2AppToBeSavedInDatabase(): void
    {
        $app_name          = 'Jenkins';
        $redirect_endpoint = 'https://example.com/redirect';
        $project           = M::mock(\Project::class);
        $new_app           = NewOAuth2App::fromAppData($app_name, $redirect_endpoint, $project);
        $this->assertSame($app_name, $new_app->getName());
        $this->assertSame($redirect_endpoint, $new_app->getRedirectEndpoint());
        $this->assertSame($project, $new_app->getProject());
    }

    public function testFromAppDataAllowsRedirectEndpointWithQuery(): void
    {
        $redirect_endpoint = 'https://example.com/redirect?key=value';
        $new_app = NewOAuth2App::fromAppData('Jenkins', $redirect_endpoint, M::mock(\Project::class));
        $this->assertSame($redirect_endpoint, $new_app->getRedirectEndpoint());
    }
}
