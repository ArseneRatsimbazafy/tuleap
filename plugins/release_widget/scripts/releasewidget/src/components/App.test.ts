/*
 * Copyright (c) Enalean, 2019 - present. All Rights Reserved.
 *
 * This file is a part of Tuleap.
 *
 * Tuleap is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * Tuleap is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Tuleap. If not, see <http://www.gnu.org/licenses/>.
 */

import { shallowMount, Wrapper } from "@vue/test-utils";
import App from "./App.vue";
import { createStoreMock } from "@tuleap-vue-components/store-wrapper-jest";
import Vue from "vue";
import GetTextPlugin from "vue-gettext";
import { StoreOptions } from "../type";

const project_id = 102;
function getPersonalWidgetInstance(store_options: StoreOptions): Wrapper<App> {
    const store = createStoreMock(store_options);
    const component_options = {
        propsData: {
            project_id
        },
        mocks: { $store: store }
    };

    Vue.use(GetTextPlugin, {
        translations: {},
        silent: true
    });

    return shallowMount(App, component_options);
}

describe("Given a release widget", () => {
    let store_options: StoreOptions & Required<Pick<StoreOptions, "getters">>;
    beforeEach(() => {
        store_options = {
            state: {
                is_loading: false
            },
            getters: {
                has_rest_error: false
            }
        };

        getPersonalWidgetInstance(store_options);
    });

    it("When there are no errors, then the widget content will be displayed", () => {
        const wrapper = getPersonalWidgetInstance(store_options);

        expect(wrapper.contains("[data-test=widget-content]")).toBeTruthy();
        expect(wrapper.contains("[data-test=show-error-message]")).toBeFalsy();
        expect(wrapper.contains("[data-test=is-loading]")).toBeFalsy();
    });

    it("When there is an error, then the widget content will not be displayed", () => {
        store_options.getters.has_rest_error = true;
        const wrapper = getPersonalWidgetInstance(store_options);

        expect(wrapper.contains("[data-test=show-error-message]")).toBeTruthy();
        expect(wrapper.contains("[data-test=widget-content]")).toBeFalsy();
        expect(wrapper.contains("[data-test=is-loading]")).toBeFalsy();
    });

    it("When it is loading rest data, then a loader will be displayed", () => {
        store_options.state.is_loading = true;
        const wrapper = getPersonalWidgetInstance(store_options);

        expect(wrapper.contains("[data-test=is-loading]")).toBeTruthy();
        expect(wrapper.contains("[data-test=widget-content]")).toBeFalsy();
        expect(wrapper.contains("[data-test=show-error-message]")).toBeFalsy();
    });

    it("When there is a rest error and it is empty, Then another message is displayed", () => {
        store_options.state.error_message = "";
        store_options.getters.has_rest_error = true;

        const wrapper = getPersonalWidgetInstance(store_options);
        expect(wrapper.find("[data-test=show-error-message]").text()).toEqual(
            "Oops, an error occurred!"
        );
    });
});
