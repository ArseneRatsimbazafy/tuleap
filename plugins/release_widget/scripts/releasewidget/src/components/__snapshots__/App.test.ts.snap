// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given a release widget When it is loading rest data, then a loader will be displayed 1`] = `
<section
  project_id="102"
>
  <div
    class="release-loader"
  />
</section>
`;

exports[`Given a release widget When the browser is IE11, Then an error message is displayed instead of the content  1`] = `
<section>
  <div
    class="tlp-alert-danger"
    data-current-language="en_US"
    data-msgid="
        The plugin \\"Release Widget\\" is not supported under IE11. Please use a more recent browser.
    "
  >
    
        The plugin "Release Widget" is not supported under IE11. Please use a more recent browser.
    
  </div>
</section>
`;

exports[`Given a release widget When there are no errors, then the widget content will be displayed 1`] = `
<section
  project_id="102"
>
  <div>
    <div
      class="project-release-widget-content"
    >
      <roadmap-section-stub />
       
      <whats-hot-section-stub />
    </div>
  </div>
</section>
`;

exports[`Given a release widget When there is a rest error and it is empty, Then another message is displayed 1`] = `
<section
  project_id="102"
>
  <div
    class="tlp-alert-danger"
  >
    
        Oops, an error occurred!
    
  </div>
</section>
`;

exports[`Given a release widget When there is an error, then the widget content will not be displayed 1`] = `
<section
  project_id="102"
>
  <div
    class="tlp-alert-danger"
  >
    
        404 Error
    
  </div>
</section>
`;
